### PR TITLE
ENH: State space: Make initialization more flexible.

### PR DESCRIPTION
**Pull request by ChadFulton**
Initializing the Kalman filter involves determining the distribution of the initial state vector, i.e. setting its mean and covariance matrices.

**Currently**, users are presented with three options:

- Known (set using known values)
- Approximate diffuse (mean = 0, covariance a diagonal matrix with large diagonal elements)
- Stationary (solve a discrete Lyapunov equation; this is currently done in Cython in the `_{{prefix}}Statespace.initialize_stationary` method)

Each of these sets the full mean and covariance matrix for the initial state vector. However, this can be inconvenient in models where e.g. some elements are stationary and some elements are non-stationary (for example, `SARIMAX` models and `UnobservedComponents` models often have this feature).

In practice, this has meant that if there are both types of states, the "Known" option has to be used, and the stationary component has to be computed manually (see e.g. `SARIMAX.initialize_state`). This is less than ideal for at least two reasons:

1. The discrete Lyapunov equation and solution has to be set up in each model, even though it's really the same each time.
2. It's slower to do this in Python space.

**In this PR**, I add the ability to specify initialization separately for separate (consecutive) blocks of the state vector, and convert all existing classes to use the new initialization system. See the docstring of the `Initialization` class for more details. This considerably simplifies initialization in `SARIMAX` and `UnobservedComponents`, and should improve performance somewhat.

**The other major benefit** of this PR is that it is one of the steps necessary to get exact diffuse initialization working.

Of course the downside is another Cython class and somehow I've managed to add 1600 lines.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
I'll volunteer to review this, probably over the weekend.

**Comment by ChadFulton** [comment]
Thanks! Also, the test failures are because I forgot to set the code so that it only uses the new initialization when "compatibility mode" (i.e. old scipy) is not on, other than that I believe it's good to go.

**Comment by ChadFulton** [comment]
After thinking about this, I think we should wait to merge this until after 0.9 for this PR when I believe we can drop the compatibility mode, so I will remove the last commit.

**Comment by ChadFulton** [comment]
@josef-pkt am I right in remembering that we are dropping support for scipy < 0.14 after 0.9?

**Comment by ChadFulton** [comment]
Whoops, compatibility mode is required for scipy < 0.16. That throws a wrench in my plans.

**Comment by ChadFulton** [comment]
@josef-pkt am I right in hoping that we are dropping support for scipy < 0.16 after 0.9?

**Comment by josef-pkt** [comment]
@ChadFulton I have no problems with dropping scipy < 0.16.
Given our slower release cycle, we lag quite far behind in what versions we support. If there is a large enough benefit, then I think we should raise the minimum requirement faster than in the past. Even in Debian we seem to have more problems keeping up with changes in pandas, scipy and numpy rather than having problems with older versions of the dependencies.

However, I don't see a strong reason to raise minimum versions fast if there is essentially no benefit for us. (E.g. personally I update my python environment maybe every two years, or on average once a year if I have computer problems.)
Aside: changes in scipy.optimize affect mostly the user with problems or availability of optimizers depending on version. The latest scipy version usually works better but not always. For me it was maybe a loss of one or two days in getting unit tests to pass across all scipy versions on Travis in the recent count model optimization refactoring. Unfortunately, some fragility in fmin_bfgs is in recent versions of scipy that we cannot avoid yet with raising the minimum requirement, e.g. to >= 0.18))

**Comment by ChadFulton** [comment]
> @ChadFulton I have no problems with dropping scipy < 0.16.

That would be great from my perspective! To be fair, this PR is not essential in any way, so we don't have to push it if there are competing reasons not to.

> However, I don't see a strong reason to raise minimum versions fast if there is essentially no benefit for us. (E.g. personally I update my python environment maybe every two years, or on average once a year if I have computer problems.)

I usually don't have strong opinions on this either. I'm mostly on Py2.7 still :|

**Comment by josef-pkt** [comment]
The main advantage of dropping older scipy version is that you can get rid of the compatibility part in statespace which should make your maintenance and development easier, independent of specific PRs.

File: statsmodels/tsa/statespace/_initialization.pxd [hunk 1]
```diff
@@ -0,0 +1,111 @@
+#cython: boundscheck=False
+#cython: wraparound=False
+#cython: cdivision=False
+"""
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/_initialization.pxd]: Does everything here need to be exposed externally?

File: statsmodels/tsa/statespace/_initialization.pxd [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                         np.float32_t [::1, :] initial_stationary_state_cov,
+                         int complex_step=*) except 1
+    cdef int clear_constant(self, int offset, np.float32_t [:] initial_state_mean) except 1
+    cdef int clear_cov(self, int offset, np.float32_t [::1, :] cov) except 1
+    cdef int initialize_known_constant(self, int offset, np.float32_t [:] initial_state_mean) except 1
+    cdef int initialize_known_stationary_cov(self, int offset, np.float32_t [::1, :] initial_stationary_state_cov) except 1
+    cdef int initialize_diffuse(self, int offset, np.float32_t [::1, :] initial_diffuse_state_cov) except 1
+    cdef int initialize_approximate_diffuse(self, int offset, np.float32_t [::1, :] initial_stationary_state_cov) except 1
+    cdef int initialize_stationary_constant(self, int offset, sStatespace model, np.float32_t [:] initial_state_mean, int complex_step=*) except 1
+    cdef int initialize_stationary_stationary_cov(self, int offset, sStatespace model, np.float32_t [::1, :] initial_stationary_state_cov, int complex_step=*) except 1
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/_initialization.pxd]: Any particular reason you use 1 as the except val?  IIUC -1 is standard.

File: statsmodels/tsa/statespace/_initialization.pyx.in [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+# Typical imports
+import numpy as np
+import warnings
+cimport numpy as np
+cimport cython
+
+np.import_array()
+
+from statsmodels.src.math cimport *
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/_initialization.pyx.in]: Maintenance would be much easier if imports were explicit instead of "cimport *".

File: statsmodels/tsa/statespace/_initialization.pyx.in [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        self.approximate_diffuse_variance = approximate_diffuse_variance
+
+        # Validate
+        tools.validate_vector_shape('known constant', &self.constant.shape[0], self.k_states, None)
+        tools.validate_matrix_shape('known covariance', &self.stationary_cov.shape[0], self.k_states, self.k_states, None)
+
+        # Internal temporary matrices
+        dim2[0] = self.k_states; dim2[1] = self.k_states;
+        self._tmp_transition = np.PyArray_ZEROS(2, dim2, {{typenum}}, FORTRAN)
+        dim2[0] = self.k_states; dim2[1] = self.k_states;
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/_initialization.pyx.in]: flake8; please split up; easier to read;

File: statsmodels/tsa/statespace/_initialization.pyx.in [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+                                 {{cython_type}} [::1, :] initial_diffuse_state_cov) except 1:
+        cdef int i
+        for i in range(offset, offset + self.k_states):
+            initial_diffuse_state_cov[i, i] = 1
+
+        return 0
+
+    cdef int initialize_approximate_diffuse(self, int offset,
+                                             {{cython_type}} [::1, :] initial_stationary_state_cov) except 1:
+        cdef int i
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/_initialization.pyx.in]: My understanding is that `Py_ssize_t` is preferred to `int` for this.  You can get away without importing it explicitly or `from cython cimport Py_ssize_t`.

File: statsmodels/tsa/statespace/initialization.py [hunk 6]
```diff
@@ -0,0 +1,607 @@
+"""
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/initialization.py]: Is this just a a non-cython implementation of `_initialization.pyx`?

File: statsmodels/tsa/statespace/_initialization.pxd (hunk 2 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/_initialization.pxd]: No reason, simply that exceptions require a return value. I guess the Cython docs have except -1, but since the other SS code has return 1 I suppose that we can keep it as is?

File: statsmodels/tsa/statespace/_initialization.pyx.in (hunk 3 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/_initialization.pyx.in]: Good point, this file doesn't need these imports.

File: statsmodels/tsa/statespace/_initialization.pyx.in (hunk 4 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/_initialization.pyx.in]: Good idea.

File: statsmodels/tsa/statespace/_initialization.pyx.in (hunk 5 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/_initialization.pyx.in]: That's very interesting, I hadn't seen that before. @bashtage could this be related to the seg faults on appveyor?

File: statsmodels/tsa/statespace/initialization.py (hunk 6 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/initialization.py]: No this is a wrapper. I'll explain more in the main comments.

**Comment by ChadFulton** [comment]
The main idea here is that if your state space model has separate groups of states, then your initial state covariance matrix will likely be block diagonal. Since the groups of states are unrelated, their initializations may be different.

For example, in an ARMAX model in which the regression coefficients are part of the state vector, the regression coefficients are given a diffuse initialization whereas the ARMA process is given a stationary initialization.

The basic idea in the code is that any group of states can either be initialized all together or can be broken up into subgroups.

Starting with the entire set of states, one could initialize it altogether (e.g. an ARMA model would just be initialized as stationary) or one could set up a few groups (e.g. the ARMAX model would have a diffuse group and a stationary group). Then, each of those groups could be initialized altogether, or they could be set up to themselves contain subgroups, and so forth.

The `Initialization` class lets you describe the initialization for a group of states, and so every group has a corresponding `Initialization` object. If the group is initialized altogether, then the `Initialization` object would have a defined `self.initialization_type`. If the group contains `n` subgroups, then the `Initialization` object would contain `n` child `Initialization` objects. Eventually, the bottom level groups have defined `self.initialization_type` and do not contain children.

Each `Initialization` object has a corresponding `_{{prefix}}Initialization`` Cython object, of the appropriate numeric type. The Cython objects actually do the work setting the block diagonal matrix elements. This work is requested by the `initialize` method of the `_{{prefix}}Representation` class, which iterates through all of the groups, subgroups, etc., and maintains an offset to figure out which diagonal block needs to be set for each group, this is used to give each block's `_{{prefix}}Initialization` object the correct location in memory.

File: statsmodels/tsa/statespace/_initialization.pxd (hunk 2 above)
Comment (jbrockmendel) [reply statsmodels/tsa/statespace/_initialization.pxd]: It looks like most of these are returning 0 unconditionally.  Getting rid of the return value/declaration would obviate the need to have the `except 1` clause at all.  The -1 seems to be a default elsewhere (namely pandas), but this is definitely not a big deal.

**Comment by jbrockmendel** [comment]
Makes sense.  Style comments aside, this looks good on a first pass.  I'll look over the non-cython file and do a second pass in the next couple days.

**Comment by jbrockmendel** [comment]
Should I hold off on reviewing until Travis errors are fixed?

**Comment by ChadFulton** [comment]
> Should I hold off on reviewing until Travis errors are fixed?

Travis errors won't be fixed, since they are all (I think) from scipy < 0.16 where this explicitly fails.

**Comment by ChadFulton** [comment]
> The main advantage of dropping older scipy version is that you can get rid of the compatibility part in statespace which should make your maintenance and development easier, independent of specific PRs.

By the way, I definitely agree with this. Looking forward to getting rid of that code!

File: statsmodels/tsa/statespace/tools.py [hunk 7]
```diff
@@ -56,10 +57,16 @@ def set_mode(compatibility=None):
     # Initialize the appropriate mode
     if not compatibility:
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/tools.py]: was it determined that compatibility is going away soon?

File: statsmodels/tsa/statespace/tests/test_representation.py [hunk 8]
```diff
@@ -666,8 +666,8 @@ def test_initialization():
     initial_state = np.zeros(2,) + 1.5
     initial_state_cov = np.eye(2) * 3.
     mod.initialize_known(initial_state, initial_state_cov)
-    assert_equal(mod._initial_state.sum(), 3)
-    assert_equal(mod._initial_state_cov.diagonal().sum(), 6)
+    assert_equal(mod.initialization.constant.sum(), 3)
+    assert_equal(mod.initialization.stationary_cov.diagonal().sum(), 6)
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/tests/test_representation.py]: The attributes being replaced are private.  Is `initialization` intended to be used directly?

File: statsmodels/tsa/statespace/_initialization.pyx.in [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        asum = blas.dzasum(&model.k_states, &model.state_intercept[0, 0], &inc)
+        {{endif}}
+
+        # If the state intercept is non-zero, compute the mean
+        if asum > tol:
+            dim2[0] = self.k_states
+            dim2[1] = self.k_states
+            ipiv = np.PyArray_ZEROS(2, dim2, np.NPY_INT32, FORTRAN)
+
+            # Create T - I
```
Comment (jbrockmendel) [inline statsmodels/tsa/statespace/_initialization.pyx.in]: In the current version of representation.pyx this is `# I - T`.  Why the change?

File: statsmodels/tsa/statespace/tools.py (hunk 7 above)
Comment (josef-pkt) [reply statsmodels/tsa/statespace/tools.py]: If this refers to the scipy version, then yes.
I think we should drop scipy < 0.16 right after the release of 0.9.

Comment (josef-pkt) [reply statsmodels/tsa/statespace/tools.py]: or we can even drop more if needed.

File: statsmodels/tsa/statespace/_initialization.pyx.in (hunk 9 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/_initialization.pyx.in]: I guess it's just easier to do using the BLAS calls that I know (of course I'm not ruling out that there may be a better way to do it altogether): subtracting 1 from the diagonal elements and then scaling the whole thing by -1 is faster than constructing an identity matrix and subtracting the T matrix from it.

File: statsmodels/tsa/statespace/_initialization.pxd (hunk 1 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/_initialization.pxd]: I missed responding to this earlier, but I'm not sure what you mean by this - that the classes should be prefixed by underscores?

File: statsmodels/tsa/statespace/_initialization.pxd (hunk 2 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/_initialization.pxd]: Yes, that's definitely doable. I guess I vaguely thought maybe returning a fixed C value would be faster than the default (which is I guess a Python None object, maybe?), but I have no real reason to believe that.

File: statsmodels/tsa/statespace/tests/test_representation.py (hunk 8 above)
Comment (ChadFulton) [reply statsmodels/tsa/statespace/tests/test_representation.py]: Yes, these may be created and manipulated by end-users, even separate from model classes (e.g. in principle someone might want to construct their own initialization even for an SARIMAX class, if e.g. they had some kind of prior).

File: statsmodels/tsa/statespace/tools.py (hunk 7 above)
Comment (jbrockmendel) [reply statsmodels/tsa/statespace/tools.py]: Should I hold off on further review until some upcoming set of edits?

File: statsmodels/tsa/statespace/_initialization.pyx.in (hunk 5 above)
Comment (jbrockmendel) [reply statsmodels/tsa/statespace/_initialization.pyx.in]: I'd be surprised.  AFAIK Py_ssize_t is basically an unsigned int.  But its the suggested usage for scalar indexing.

File: statsmodels/tsa/statespace/_initialization.pxd (hunk 2 above)
Comment (jbrockmendel) [reply statsmodels/tsa/statespace/_initialization.pxd]: I think `cdef void foo` might achieve both goals.

File: statsmodels/tsa/statespace/_initialization.pxd (hunk 1 above)
Comment (jbrockmendel) [reply statsmodels/tsa/statespace/_initialization.pxd]: One of the purposes of the pxd file is to make it expose things to be `cimport`ed.  (The other big one is for type declarations, but since you already have those in the pyx file that is unnecessary).  So anything that is not needed outside of statespace._intialization can be excluded from the pxd file.

The main benefit of this is for the reviewer: if it is easy to check that a function/class is only used internally, it becomes much easier to track reason about its use cases.

**Comment by bashtage** [comment]
> That's very interesting, I hadn't seen that before. @bashtage could this be related to the seg faults on appveyor?

Unlikely I caught one locally with a debug build. The segfault is a 0xc5 where something is trying to read self.nobs after self has been garbage collected.  Very strange.

**Comment by ChadFulton** [comment]
superseded by #4418
